source/common/network/connection_impl.h

```diff
@@ -81,8 +82,10 @@ class ConnectionImpl : public virtual Connection,
   const Address::InstanceConstSharedPtr& remoteAddress() const override { return remote_address_; }
   const Address::InstanceConstSharedPtr& localAddress() const override { return local_address_; }
   void setConnectionStats(const ConnectionStats& stats) override;
-  Ssl::Connection* ssl() override { return nullptr; }
-  const Ssl::Connection* ssl() const override { return nullptr; }
+  Ssl::Connection* ssl() override { return dynamic_cast<Ssl::SslSocket*>(transport_socket_.get()); }
```

Comment:
this dynamic cast seems a bit fragile and magical. Should transport socket itself have an `ssl()` method that can return nullptr or not? Potentially it can just return `this` in many cases.

Reply:
Yeah, I think I'd requested the same in the original PR.  Let's do it!

My one concern (which we don't need to deal with here esp. if it causes lots of name churn, but I want folks thinking about) is that this is ssl-specific.  We had similar APIs in our proxy and had to rewrite them all for QUIC crypto which had overlapping for not identical concepts.  The more we can keep naming and APIs crypto-impl-neutral as we refactor the better.

Reply:
Done for ssl() in transport socket.